processor/groupbytraceprocessor/processor.go

```diff
@@ -254,32 +217,25 @@ func (sp *groupByTraceProcessor) markAsReleased(traceID pdata.TraceID) error {
 	// signal that the trace is ready to be released
 	sp.logger.Debug("trace marked as released", zap.Stringer("traceID", traceID))
 
-	sp.cancelLock.RLock()
-	if !sp.cancelReleases {
-		sp.released <- trace
-		sp.removed <- traceID
-	}
-	sp.cancelLock.RUnlock()
+	// atomically fire the two events, so that a concurrent shutdown won't leave
+	// an orphaned trace in the storage
+	sp.eventMachine.fire(event{
+		typ:     traceReleased,
+		payload: trace,
+	}, event{
+		typ:     traceRemoved,
```

Comment:
`traceRemoved` is now the only event type emitted in 2 places.
It works fine, but I was wondering if we can group them differently.

The real events which trigger a change are `received` (in), `expired` and `evicted` (out). Others are parts of their processing:
- "released" is part of `onExpired`.
- "removed" is part of `onExpired` or `onEvicted`

Would there be a negative performance impact if we grouped them this way?
If you don't like this idea, feel free to leave it as it is.

Reply:
Actually, "evicted" might also be part of `onReceived`, so there would be just 2 event types - `received` and `expired`, but that might be an overkill.